### PR TITLE
support weighted average aggregation function

### PR DESCRIPTION
**Pull request by gs-gunjan**
#### What type of PR is this?
- Improvement

#### What does this PR do / why is it needed ?
Supporting weighted average aggregation function

#### Which issue(s) this PR fixes:
NA

#### Other notes for reviewers:
NA

#### Does this PR introduce a user-facing change?
NA

### CODE REVIEW

**Review by akphi — APPROVED** [review]

**Comment by akphi** [comment]
@gs-gunjan looks good to me, do you want to merge this?

File: legend-engine-xts-relationalStore/legend-engine-xt-relationalStore-generation/legend-engine-xt-relationalStore-pure/src/main/resources/core_relational/relational/pureToSQLQuery/pureToSQLQuery.pure [hunk 1]
```diff
@@ -4822,10 +4827,17 @@ function meta::relational::functions::pureToSqlQuery::processAggFn(v:ValueSpecif
                       pair($f.func == or_Boolean_MANY__Boolean_1_ && $singleArgument, 'boolor'),
                       pair(true, $fn)
                     ]->filter(p | $p.first)->first()->toOne().second;
-                    let aggFunc = newDynaFunction(
+
+                    let aggFunc = if ($f.func == wavg_WavgRowMapper_MANY__Float_1_,
+                    | let parameters = $f.parametersValues->map(p|$p->processAggFn($params, $state));
+                      if ($parameters->size() == 1 && $parameters->toOne()->instanceOf(DynaFunction) && $parameters->cast(@DynaFunction).parameters->size()== 2 && $parameters->cast(@DynaFunction).name == 'wavgRowMapper',
```
Comment (gs-ssh16) [inline legend-engine-xts-relationalStore/legend-engine-xt-relationalStore-generation/legend-engine-xt-relationalStore-pure/src/main/resources/core_relational/relational/pureToSQLQuery/pureToSQLQuery.pure:4833]: When will this condition evaluate to false and does a test case cover it?

Comment (gs-gunjan) [reply legend-engine-xts-relationalStore/legend-engine-xt-relationalStore-generation/legend-engine-xt-relationalStore-pure/src/main/resources/core_relational/relational/pureToSQLQuery/pureToSQLQuery.pure:4833]: yes its covered in the test written, that else part is called when being called by old TDS flow

File: legend-engine-core/legend-engine-core-pure/legend-engine-pure-code-functions-standard/legend-engine-pure-functions-standard-pure/src/main/resources/core_functions_standard/math/aggregator/wavg.pure [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+
+function meta::pure::functions::math::wavgUtility::wavgRowMapper(quantity:Number[0..1], weight:Number[0..1]):meta::pure::functions::math::wavgUtility::WavgRowMapper[1]
+{
+  ^meta::pure::functions::math::wavgUtility::WavgRowMapper(
+    quantity= $quantity,
+    weight= $weight
+  )
+}
+
+function <<PCT.test>> meta::pure::functions::math::tests::wavg::testSimpleGroupByWavg<T|m>(f:Function<{Function<{->T[m]}>[1]->T[m]}>[1]):Boolean[1]
```
[the remark this answers is not in the document]
Comment (gs-gunjan) [reply legend-engine-core/legend-engine-core-pure/legend-engine-pure-code-functions-standard/legend-engine-pure-functions-standard-pure/src/main/resources/core_functions_standard/math/aggregator/wavg.pure:54]: added tests

File: legend-engine-core/legend-engine-core-pure/legend-engine-pure-code-functions-standard/legend-engine-pure-functions-standard-pure/src/main/resources/core_functions_standard/math/aggregator/wavg.pure [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import meta::pure::test::pct::*;
+
+function <<PCT.function>> meta::pure::functions::math::wavg(numbers:Number[*], weights:Number[*]):Float[1]
+{
+    if($numbers->isEmpty() || $weights->isEmpty() || $numbers->size() != $weights->size(),
+      |fail('Error: Mean of an empty set.');0.0;,
```
[the remark this answers is not in the document]
Comment (gs-gunjan) [reply legend-engine-core/legend-engine-core-pure/legend-engine-pure-code-functions-standard/legend-engine-pure-functions-standard-pure/src/main/resources/core_functions_standard/math/aggregator/wavg.pure:26]: error added specifically for this case

File: legend-engine-core/legend-engine-core-pure/legend-engine-pure-code-functions-standard/legend-engine-pure-functions-standard-pure/src/main/resources/core_functions_standard/math/aggregator/wavg.pure [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+      | let weightedQuantitySum = $numbers->size()->range()->fold({i, y| $y + toFloat($numbers->at($i) * $weights->at($i))}, 0.0);
```
[the remark this answers is not in the document]
Comment (gs-gunjan) [reply legend-engine-core/legend-engine-core-pure/legend-engine-pure-code-functions-standard/legend-engine-pure-functions-standard-pure/src/main/resources/core_functions_standard/math/aggregator/wavg.pure]: changed it

**Review by aziemchawdhary-gs — APPROVED** [review]

**Pull request merged**